storybookjs/storybook
# Manager: Fix useAddonState when using a setter function

Closes https://github.com/storybookjs/storybook/issues/23787

## What I did

The `useAddonState` hook allows for a setter function.
This setter function takes the existing value, and returns a new value.
However due to a race condition due to the way we set the default state (without an sync re-render) this default state can be set in the state too late.

By globalizing the addonCache and retreiving the value form the cache instead of state when `api.getAddonState` is called we ensure the correct value is always set.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

The issue here: https://github.com/storybookjs/storybook/issues/23787
... has a repro to reproduce the issue with.

- create a repro out of the download (follow some extra steps)
- build storybook in the monorepo
- copy `code/ui/manager/dist` to the `node_modules/@storybook/manager/dist` directory
- run the storybook in the repro

expect that the storybook functions correctly

- click the primary button manually
- refresh the page
- do it again
- open the interactions tab
- press the button manually
- refresh the page



### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
